File: lib/computed.ts
import { Dependency, DirtyLevels, Subscriber } from './system.js';

export function computed<T>(getter: (cachedValue?: T) => T) {
	return new Computed<T>(getter);
}

export class Computed<T = any> implements Dependency, Subscriber {
	cachedValue: T | undefined = undefined;

	// Dependency
	subs = undefined;
	subsTail = undefined;
	subVersion = -1;

	// Subscriber
	deps = undefined;
	depsTail = undefined;
	versionOrDirtyLevel = DirtyLevels.Dirty;

	constructor(
		public getter: (cachedValue?: T) => T
	) { }

	notifyLostSubs() {
		Subscriber.clearTrack(this);
		this.versionOrDirtyLevel = DirtyLevels.Dirty;
	}

	get(): T {
		Dependency.linkSubOnly(this);
		const dirtyLevel = this.versionOrDirtyLevel;
		if (dirtyLevel === DirtyLevels.MaybeDirty) {
			Subscriber.resolveMaybeDirty(this);
			if (this.versionOrDirtyLevel === DirtyLevels.Dirty) {
				return this.update();
			}
		} else if (dirtyLevel === DirtyLevels.Dirty) {
			return this.update();
		}
		return this.cachedValue!;
	}

	update() {
		const prevSub = Subscriber.startTrack(this);
		const oldValue = this.cachedValue;
		const newValue = this.getter(oldValue);
		Subscriber.endTrack(this, prevSub);
		if (oldValue !== newValue) {
			this.cachedValue = newValue;
			Dependency.propagate(this);
		}
		return newValue;
	}
}

File: src/main/kotlin/com/example/Computed.kt
package com.example

class Computed<T>(private val getter: () -> T) {

    private var cachedValue: T? = null
    private var version: Int = 0

    private var subs: MutableList<Subscriber> = mutableListOf()
    private var lastTrackedId: Int = 0

    private var deps: MutableList<Signal<*>> = mutableListOf()
    private var flags: Int = 0

    fun get(): T {
        if (flags and DIRTY != 0) {
            update()
        } else if (flags and TO_CHECK_DIRTY != 0) {
            if (checkDirty(deps)) {
                update()
            } else {
                flags = flags and TO_CHECK_DIRTY.inv()
            }
        }
        if (activeTrackId != null) {
            if (lastTrackedId != activeTrackId) {
                lastTrackedId = activeTrackId
                link(this, activeSub!!)
            }
        } else if (activeScopeTrackId != null) {
            if (lastTrackedId != activeScopeTrackId) {
                lastTrackedId = activeScopeTrackId
                link(this, activeEffectScope!!)
            }
        }
        return cachedValue!!
    }

    fun update(): Boolean {
        val prevSub = activeSub
        val prevTrackId = activeTrackId
        setActiveSub(this, nextTrackId())
        startTrack(this)
        val oldValue = cachedValue
        val newValue: T
        try {
            newValue = getter()
        } finally {
            setActiveSub(prevSub, prevTrackId)
            endTrack(this)
        }
        if (oldValue != newValue) {
            cachedValue = newValue
            version++
            return true
        }
        return false
    }

    private fun checkDirty(deps: List<Signal<*>>): Boolean {
        for (dep in deps) {
            if (dep is Computed<*>) {
                if (dep.version != version) {
                    return true
                }
                if (dep.flags and DIRTY != 0) {
                    if (dep.update()) {
                        return true
                    }
                } else if (dep.flags and TO_CHECK_DIRTY != 0) {
                    if (checkDirty(dep.deps)) {
                        if (dep.update()) {
                            return true
                        }
                    } else {
                        dep.flags = dep.flags and TO_CHECK_DIRTY.inv()
                    }
                }
            }
        }
        return false
    }

    private fun link(dep: Signal<*>, sub: Subscriber) {
        sub.deps.add(dep)
        dep.subs.add(sub)
    }

    companion object {
        const val DIRTY = 1 shl 4
        const val TO_CHECK_DIRTY = 1 shl 3
    }
}

interface Subscriber {
    val deps: MutableList<Signal<*>>
    fun notify()
}

var activeTrackId: Int? = null
var activeSub: Subscriber? = null
var activeScopeTrackId: Int? = null
var activeEffectScope: Subscriber? = null

fun setActiveSub(sub: Subscriber?, trackId: Int?) {
    activeSub = sub
    activeTrackId = trackId
}

fun nextTrackId(): Int {
    return (activeTrackId ?: 0) + 1
}

fun startTrack(sub: Subscriber) {
    sub.deps.clear()
}

fun endTrack(sub: Subscriber) {
    // No-op for now
}
